airflow/worker/task_runner_worker.py

```diff
@@ -0,0 +1,184 @@
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+
+import logging
+import os
+from datetime import datetime
+import asyncio
+from aiohttp import web
+import pendulum
+from contextlib import suppress
+from typing import Dict
+from airflow import settings
+from airflow.exceptions import AirflowException
+from airflow.models import DAG
+from airflow.models import (
+    DagBag, TaskInstance
+)
+from airflow.utils.log.logging_mixin import (LoggingMixin)
+from airflow.utils.net import get_hostname
+from airflow.utils.state import State
+from concurrent.futures import ProcessPoolExecutor
+
+app = None
+loop = None
+pool = None
+executor = None
+DAGS_FOLDER = settings.DAGS_FOLDER
+
+running_tasks_map = {}  # type: Dict[str, TaskInstance]
+
+heartbeat_loop_task = None
+num = 0
+
+
+async def heartbeat():
+    global running_tasks_map, num
+    while True:
+        num = num + 1
+        for k, v in running_tasks_map.items():
+            a = v
+            await a.heartbeat()
+        await asyncio.sleep(1)
+
+
+async def health(request):
+    name = request.rel_url.query.get("name", "Daniel")
+    return web.Response(text="Hello, {}. There have been {} heartbeats.".format(name, num))
+
+
+async def run_task(request):
+    global running_tasks_map
+    dag_id = request.rel_url.query['dag_id']
+    task_id = request.rel_url.query['task_id']
+    subdir = None
+    response = None
+    execution_date = pendulum.fromtimestamp(int(request.rel_url.query["execution_date"]))
+    log = LoggingMixin().log
+    log.info("running dag {} for task {} on date {} in subdir {}"
+             .format(dag_id, task_id, execution_date, subdir))
+    logging.shutdown()
+    key = (task_id, dag_id, execution_date).__str__()
+    try:
+        # IMPORTANT, have to use the NullPool, otherwise, each "run" command may leave
+        # behind multiple open sleeping connections while heartbeating, which could
+        # easily exceed the database connection limit when
+        # processing hundreds of simultaneous tasks.
+        settings.configure_orm(disable_connection_pool=True)
+        ti = get_task_instance(dag_id=dag_id,
+                               task_id=task_id,
+                               subdir=process_subdir(subdir),
+                               execution_date=execution_date)
+        out = run(ti)
+        running_tasks_map[key] = ti
+        run_task_instance(ti, log)
```

Comment:
`run(ti)` calls `run_task_instance()`, but we call it again here. Are we running each task twice?